lib/ex_unit/lib/ex_unit/pattern_diff.ex

```diff
@@ -0,0 +1,391 @@
+defmodule ExUnit.Pattern do
+  defstruct [:binary, :val, :match_single?, :vars, :pins]
+
+  @type t :: %__MODULE__{
+          binary: String.t(),
+          val: any(),
+          vars: map,
+          pins: [key: atom()],
+          match_single?: boolean()
+        }
+
+  def new(lh_pattern, pins, unbound_vars, match_single? \\ true)
+      when is_list(pins) and is_map(unbound_vars) do
+    %__MODULE__{
+      binary: Macro.to_string(lh_pattern),
+      val: lh_pattern,
+      pins: pins,
+      vars: unbound_vars,
+      match_single?: match_single?
+    }
+  end
+
+  def get_var({var, meta, context}) when is_atom(var) and is_atom(context) do
+    {var, meta[:counter] || context}
+  end
+end
+
+defmodule ExUnit.ContainerDiff do
+  defstruct [:type, :items]
+
+  @type t :: %__MODULE__{
+          type: :list | :map | :tuple | :struct | :when,
+          items: [t | ExUnit.KeyDiff | ExUnit.PatternDiff.t()]
+        }
+end
+
+defmodule ExUnit.WhenDiff do
+  defstruct [:op, :bindings, :result]
+
+  @type t :: %__MODULE__{
+          op: atom(),
+          bindings: Keyword.t(),
+          result: :eq | :neq
+        }
+end
+
+defmodule ExUnit.PatternDiff do
+  alias ExUnit.{ContainerDiff, WhenDiff}
+
+  defstruct [:type, :lh, :rh, :diff_result]
+
+  @type lhs :: %{
+          type: atom(),
+          ast: any()
+        }
+
+  @type t :: %__MODULE__{
+          type: :value | :key | :different | :map | :struct | :list | :tuple,
+          lh: any(),
+          rh: any(),
+          diff_result: :eq | :neq
+        }
+
+  @no_value :ex_unit_no_meaningful_value
+
+  # {:^, _, var}
+  # {:|, _, [l, r]}
+  # {:when, _, [l, r]}
+  # {:=, _, [l, r]}
+  # {:%{}, _, members}
+  # {:{}, _, members}
+  # {:_, _, _}
+  # {var, _, context}
+  # {key, val}
+  # list
+  # integer
+  # float
+  # string
+  # atom
```

Comment:
What abut binary patterns `{:<<>>, _, _}`?

Reply:
So far I have punted on those, focusing on the lists & maps, which I think covers most of what is desired.  Since the approach for those will most likely be the same as we do here, I would prefer to work out any issues with this PR before getting too deep in them?  Thoughts?

Reply:
Makes sense to punt. Doing a full diff on binaries would also be hard so it is likely that we will just compile them to something like `match?(binary_pattern, value)` and treat the whole thing as true or false.